### PR TITLE
Initial support for using a remote Postgres cluster as a backend

### PR DESCRIPTION
**Pull request by elprans**
This adds the new `--postgres-dsn` option to `edgedb-server`.  When 
specified instead of `-D`, EdgeDB will connect to the specified cluster 
address instead of using a local data directory.

Supporting this requires a bunch of changes around how roles and 
connections are treated.  Since most DBaaS restrict direct superuser 
access, and instead provide a regular role with specific privileges, the 
bootstrap process became somewhat more involved.

### CODE REVIEW

File: edb/pgsql/metaschema.py [hunk 1]
```diff
@@ -2399,9 +2399,9 @@ def _generate_role_views(schema):
             a.rolname                                   AS name,
             a.rolsuper                                  AS is_superuser,
             a.rolcanlogin                               AS allow_login,
-            a.rolpassword                               AS password
+            (d.description)->>'ph'                      AS password
```
Comment (tailhook) [inline edb/pgsql/metaschema.py]: Do you really think that compressing/obfuscating this field is justified?

File: edb/server/pgcon/pgcon.pyx [hunk 2]
```diff
@@ -949,6 +981,22 @@ cdef class PGProto:
...
 
+    cdef make_auth_password_md5_message(self, bytes salt):
+        cdef WriteBuffer msg
+
+        msg = WriteBuffer.new_message(b'p')
+
+        user = self.pgaddr.get('user') or ''
+        password = self.pgaddr.get('password') or ''
+
+        # 'md5' + md5(md5(password + username) + salt))
```
Comment (tailhook) [inline edb/server/pgcon/pgcon.pyx]: Is this encoding of a password is still considered safe? I would expect PBKDF with thousands of iterations here too. (perhaps, argon2_ffi?)

File: docs/admin/setup.rst [hunk 3]
```diff
@@ -68,3 +68,31 @@ method with the ``trust`` method:
 .. code-block:: bash
 
     $ edgedb --admin -h </data/dir> configure insert auth --method=trust
+
+
+Using remote PostgreSQL cluster as a backend
```
Comment (tailhook) [inline docs/admin/setup.rst]: Please, document environment and pgpass file

File: edb/server/pgcon/pgcon.pyx (hunk 2 above)
Comment (elprans) [reply edb/server/pgcon/pgcon.pyx]: This is what RDS Postgres advertises.  Ideally, they'd need to configure it to use SCRAM, but they didn't.

File: edb/pgsql/metaschema.py (hunk 1 above)
Comment (elprans) [reply edb/pgsql/metaschema.py]: Not really.  Security through obscurity does not work.  We have to rely on a) That access to `pg_shdescription` is sufficiently restricted; and b) That SCRAM makes brute-force attacks infeasible.

Comment (tailhook) [reply edb/pgsql/metaschema.py]: Yes. I mean, rename it to `password_hash` or something :)

File: edb/server/pgcon/pgcon.pyx (hunk 2 above)
Comment (tailhook) [reply edb/server/pgcon/pgcon.pyx]: Ah, sorry, I've misread this as our method for storing passwords.

File: docs/admin/setup.rst (hunk 3 above)
Comment (elprans) [reply docs/admin/setup.rst]: Done.

File: edb/pgsql/metaschema.py (hunk 1 above)
Comment (elprans) [reply edb/pgsql/metaschema.py]: Ah, OK. Done.

Comment (elprans) [reply edb/pgsql/metaschema.py]: Ah, OK. Done.

**Pull request merged**